File: cpp/1805. 字符串中不同整数的数目.cpp
class Solution {
public:
    int numDifferentIntegers(string word) {
        int n=word.size();
        set<string> st;
        for(int i=0;i<n;){
            string t="";
            while(word[i]>='0'&&word[i]<='9'){
                t+=word[i];
                i++;
            }
            if(t!="") {
                while(t[0]=='0'){
                    t.erase(t.begin());
                }
                st.insert(t);
            }
            i++;
        }
        return st.size();
    }
};

File: 刷题/golang/1805. 字符串中不同整数的数目.go
func numDifferentIntegers(word string) int {
    s := map[string]bool{}
    n := len(word)
    p1 := 0
    for {
        for p1 < n && !unicode.IsDigit(rune(word[p1])) {
            p1++
        }
        if p1 == n {
            break
        }
        p2 := p1
        for p2 < n && unicode.IsDigit(rune(word[p2])) {
            p2++
        }
        for p2-p1 > 1 && word[p1] == '0' { // 去除前导 0
            p1++
        }
        s[word[p1:p2]] = true
        p1 = p2
    }
    return len(s)
}